src/rdsk.c

```diff
@@ -280,54 +266,39 @@ struct RC_PROFILE *search_cache_targets(char *return_message)
 		print_error(msg, return_message, __func__, strerror(errno));
 		return NULL;
 	}
-	if ((num2 = scandir(SYS_BLOCK, &nodes, scandir_filter_dm, NULL)) < 0) {
-		msg = ERR_SCANDIR;
-		print_error(msg, return_message, __func__, strerror(errno));
-		list = clean_scandir(list, num);
-		return NULL;
-	}
 
-	for (;n < num; n++) {
+	for (n = 0; n < num; n++) {
 		prof = calloc(1, sizeof(struct RC_PROFILE));
 		if (prof == NULL) {
 			msg = ERR_CALLOC;
-			print_error(msg, return_message, __func__, strerror(errno));
-			list = clean_scandir(list, num);
-			nodes = clean_scandir(nodes, num2);
-			free_linked_lists(cache_head, NULL, NULL);
-			return NULL;
+			goto search_cache_targets_error;
 		}
 		strcpy(prof->device, list[n]->d_name);
+		sprintf(file, "%s/%s", DEV_MAPPER, list[n]->d_name);
+		readlink(file, link, sizeof(link));
```

Comment:
`readlink` call needs a check, its result must not be `-1`. On error, `errno` is set.

Reply:
Another good catch.